addons/mail/models/mail_message.py

```diff
@@ -390,7 +390,8 @@ def filter_notification(notif):
             for notification in message.notification_ids.filtered(filter_notification):
                 customer_email_data.append((partner_tree[notification.res_partner_id.id][0], partner_tree[notification.res_partner_id.id][1], notification.email_status))
 
-            main_attachment = message.model and message.res_id and self.env[message.model].search([('id', '=',message.res_id)]) and getattr(self.env[message.model].browse(message.res_id), 'message_main_attachment_id')
+            has_access_to_model = message.model and self.env[message.model].check_access_rights('read', raise_exception=False)
+            main_attachment = has_access_to_model and message.res_id and self.env[message.model].search([('id', '=',message.res_id)]) and getattr(self.env[message.model].browse(message.res_id), 'message_main_attachment_id')
```

Comment:
for speed you could use `search_count` in `self.env[message.model].search([('id', '=',message.res_id)])`

you don't need check `access_right` if there are no `res_id`. 
Load one time the `env` of the `model` (3 times before), it is better for performance.

```python
main_attachment=False
if message.model and message.res_id:
    CoModel = self.env[message.model]
    has_access_to_model = CoModel.check_access_rights('read', raise_exception=False)
    if has_access_to_model and CoModel.search_count([('id', '=',message.res_id)]) 
         main_attachment = getattr(CoModel.browse(message.res_id), 'message_main_attachment_id')

```